### PR TITLE
Fixed #28216 -- Added next_page/get_default_redirect_url() to LoginView.

### PR DESCRIPTION
**Pull request by ThinkChaos**
[Django Ticket](https://code.djangoproject.com/ticket/28216)

### CODE REVIEW

File: docs/releases/1.11.3.txt [hunk 1]
```diff
@@ -6,6 +6,11 @@ Django 1.11.3 release notes
 
 Django 1.11.3 fixes several bugs in 1.11.2.
 
+Minor feature
```
Comment (timgraham) [inline docs/releases/1.11.3.txt]: This won't be backported to 1.11 per our [supported versions policy](https://docs.djangoproject.com/en/dev/internals/release-process/#supported-versions). This should be in the 2.0 release notes.

File: docs/topics/auth/default.txt [hunk 2]
```diff
@@ -945,7 +945,7 @@ All authentication views
 This is a list with all the views ``django.contrib.auth`` provides. For
 implementation details see :ref:`using-the-views`.
 
-.. function:: login(request, template_name=`registration/login.html`, redirect_field_name='next', authentication_form=AuthenticationForm, extra_context=None, redirect_authenticated_user=False)
+.. function:: login(request, next_page=None, template_name=`registration/login.html`, redirect_field_name='next', authentication_form=AuthenticationForm, extra_context=None, redirect_authenticated_user=False)
```
Comment (timgraham) [inline docs/topics/auth/default.txt]: The function-based view isn't modified (and shouldn't be).

**Comment by jdufresne** [comment]
I have a use case such that the default login redirect depends on a number of runtime values. Mostly global settings and per-user settings. Rather than a class attribute, maybe this hook should be implemented as a method. A method would provide full flexibility for calculating the default redirect page for a custom LoginView. I was thinking the following set of functions would easily allow my usecase without requiring me to completely rewrite `get_success_url()` (as I'm doing now).

```python
    def get_success_url(self):
        return self.get_redirect_url() or self.get_default_redirect_url()

    def get_redirect_url(self):
        """Return the user-originating redirect URL if it's safe."""
        redirect_to = self.request.POST.get(
            self.redirect_field_name,
            self.request.GET.get(self.redirect_field_name, '')
        )
        url_is_safe = is_safe_url(
            url=redirect_to,
            allowed_hosts=self.get_success_url_allowed_hosts(),
            require_https=self.request.is_secure(),
        )
        return redirect_to if url_is_safe else ''

    def get_default_redirect_url(self):
        return resolve_url(settings.LOGIN_REDIRECT_URL)
```

For my use case, I'd be only overriding `get_default_redirect_url()`.

If you stick with the class attribute, I suppose I could potentially solve this by using `@property`.

**Comment by timgraham** [comment]
@ThinkChaos, are you able to update this as per the comments?

File: docs/topics/auth/default.txt (hunk 2 above)
Comment (ThinkChaos) [inline docs/topics/auth/default.txt]: The login function takes `**kwargs`, and forwards it to `LoginView`, so without changing the function, we get the extra functionality.
Also, I mirrored the logout function's docs ([line 1100](https://github.com/ThinkChaos/django/blob/4034a5cbef7e9b6afb2f151ec2201ab5afbeeaf3/docs/topics/auth/default.txt#L1100)), do you still want me to change this?

**Comment by ThinkChaos** [comment]
@jdufresne Yeah I agree, I wanted the same at some point, but I believe it's kind of outside the scope of this PR as it would be nice to change that on all similar view classes.

**Comment by jdufresne** [comment]
> but I believe it's kind of outside the scope of this PR

It is changing the approach of using a class attribute passed to `resolve_url()` to using a method that returns a URL. I see it as an alternative implementation that achieves your use case as well as mine. In that sense, I see it in scope of achieve the larger goals of the feature.

> as it would be nice to change that on all similar view classes

Which other classes are you referring?

**Comment by ThinkChaos** [comment]
> Which other classes are you referring?

In `django.views.generic.edit`, I just checked, and it doesn't seem as useful there.
I just pushed a commit with the change.

File: docs/topics/auth/default.txt (hunk 2 above)
Comment (timgraham) [reply docs/topics/auth/default.txt]: It's no longer the case after 2dacc2ccd9593678f19b61ecfe1bc6ee826c2a88, so yes, please remove it.

**Comment by jdufresne** [comment]
The logic changes in `views.py` look good to me and will solve my use case.

**Comment by timgraham** [comment]
Is the `get_default_redirect_url()` method needed? If so, I guess it should be documented and tested for overriding.

**Comment by jdufresne** [comment]
> Is the get_default_redirect_url() method needed?

It was added at my request in comment https://github.com/django/django/pull/8630#issuecomment-308436367

It makes calculating the default redirect URL at runtime easier based on the user signing in (think admin vs normal user), runtime settings from the database, etc.

> If so, I guess it should be documented and tested for overriding.

I agree. That would be a nice improvement.

**Comment by ThinkChaos** [comment]
I rolled back the changes I had made to `login` docs.

Where should the docs for `get_default_redirect_url` go?  
I was thinking with the `LoginView` class in `docs/topics/auth/default.txt`, does that sound good?

**Comment by LuRsT** [comment]
@ThinkChaos, do you still need help with where to place documentation? If you still want to work on this, just write them where you think it's sensible and someone will gently ask you to change it if needed, no need to ask for permission :sparkles:

**Comment by ThinkChaos** [comment]
Rebased and added the `get_default_redirect_url` docs.

**Comment by felixxm** [comment]
@ThinkChaos Thanks :+1: I rebased, added extra tests, and pushed minor edits. I'm going to add release notes and polish docs after #13988.

**Comment by felixxm** [comment]
@ThinkChaos Welcome aboard :boat:

File: docs/releases/4.0.txt [hunk 3]
```diff
@@ -47,6 +47,12 @@ Minor features
 * The default iteration count for the PBKDF2 password hasher is increased from
   260,000 to 320,000.
 
+* The new
+  :attr:`LoginView.next_page <django.contrib.auth.views.LoginView.next_page>`
+  attribute and
+  :meth:`~django.contrib.auth.views.LoginView.get_default_redirect_url` method
+  allow customizing the redirect to after login.
```
Comment (smithdc1) [inline docs/releases/4.0.txt:54]: Nit pick. I keep tripping up on the 'to'. I think it can be chopped?

Comment (felixxm) [reply docs/releases/4.0.txt]: Updated.

**Comment by ThinkChaos** [comment]
@felixxm Thanks for picking this up!

**Pull request merged**